Lexer

svmnotn opened the pull request:
This pr is not finished yet, check that it is satisfactory and recommend any more tests that could be added.

svmnotn committed:
feat: add location to token.

cksac commented:
I think Int and Float token should parse into actual type at later stage, The built-in Scalar in GraphQL is i32 and f64, and input Int value is valid or not depends on which type it refer to, e.g. build-in Int or custom scalar Int8.

svmnotn committed:
fix numbers, and add symbol testing

svmnotn commented:
I think that covers all the symbols

cksac commented on src/language/lexer.rs:
```diff
@@ -372,24 +372,44 @@ mod tests {
+    test_next_token(&mut lexer, Token::RightBrace(24,24));
+    test_next_token(&mut lexer, Token::RightParen(26,26));
+    test_next_token(&mut lexer, Token::Spread(33,35));
+    test_next_token(&mut lexer, Token::Eof(33,36));
```
The location for Eof token seems a bit odd.

svmnotn replied:
Yeah, I found the locations by reading the assertion errors, so the lexer is probably doing something weird there.

cksac commented on src/language/lexer.rs:
```diff
@@ -372,24 +372,44 @@ mod tests {
+  #[test]
+  fn test_symbol() {
+    let mut lexer = Lexer::new("@ !  :$ =    [  {   (|] } )      ...");
+    test_next_token(&mut lexer, Token::At(0,0));
```
I would expect I can use the location information to get back the string captured in token from lexer input. e.g. lexer.input[lo..hi] = "@", but currently it does not.

calebmer commented on src/language/lexer.rs:
```diff
@@ -2,26 +2,26 @@
+  Pipe(usize, usize),
+  Name(usize, usize, &'a str),
+  Int(usize, usize, &'a str),
+  Float(usize, usize, &'a str),
```
As a note, the spec models tokens differently where `Bang`, `Dollar`, etc. are modeled by a `Punctuator` type. So the types might instead look like this:

``` rust
pub enum Token<'a> {
  Eof(…),
  Punctuator(usize, Punctuator),
  IntValue(…),
  FloatValue(…),
  StringValue(…)
}

pub enum Punctuator {
  Bang,
  Dollar,
  LeftParen,
  …
}
```

I think this might make code cleaner later when trying to `match` against tokens to parse into grammar types. Where there is only one `usize` because we can guarantee there will only be a single character (might be wrong on the one `usize` there, I have no experience writing lexer/parsers 😊).

**Edit**: Two `usize`s are required, there is no guarantee that a punctuator is a single character as `...` is a punctuator.

At the very least I think you should change `Int`, `Float`, and `String` to `IntValue`, `FloatValue`, and `StringValue` per the [specification](http://facebook.github.io/graphql/#sec-Appendix-Grammar-Summary.Lexical-Tokens).

cksac replied:
Agree to rename Int, Float, and String to IntValue, FloatValue, and StringValue.
for the Punctuator token, how about Punctuator(&'a str, usize, usize) ?
later it can pattern match like Token::Punctuator("...", _, _), not sure this is slow compare to 
Token::Punctuator(Punctuator::Spread, _, _), but this looks odd to me
I also prefer put the location information at the end.

calebmer replied:
Location information at the end sounds like a great idea, but as for punctuators it isn't so much a matter of performance to me, but rather a matter of I just really like enums 😊

Type safety and pattern exhaustion are super powerful tools in Rust (coming from a JavaScript/Haskell background), so we should really try to use them wherever possible. I agree it doesn't look the prettiest, but that's what macros/compiler plugins are for 😉

The [Iron web framework](http://ironframework.io/) went this way with headers. For example a `Content-Type` header looks like: `ContentType(Mime(TopLevel::Text, SubLevel::Html, vec![(Attr::Charset, Value::Utf8)]))`. Their justification (which I agree with even if I hate writing it) can be found [here](http://ironframework.io/doc/iron/headers/index.html#why-typed).

To further alleviate ugliness, we could also do what the [standard library does with the `Result` and `Option` enums](http://doc.rust-lang.org/std/prelude/index.html) in that we write `pub use Puncuator::*`. This changes `Token::Punctuator(Puncuator::Spread, _, _)` to just `Token::Puncuator(Spread, _, _)` a much cleaner and agreeable syntax.

Thoughts?

calebmer commented on src/language/lexer.rs:
```diff
@@ -2,26 +2,26 @@
 use std::iter::Peekable;
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq)]
```
Style thing: I typically prefer `#[derive(PartialEq, Debug)]` as it better models the importance of the derivations. Generally `Debug` is only used when tests fail whereas `PartialEq` is important in non-test code.

calebmer commented:
@svmnotn you got all the tests 👍

svmnotn committed:
style: PartialEq before Debug

svmnotn committed:
feat: add Punctuator

svmnotn committed:
feat: use new Punctuator enum, and fix tests

svmnotn committed:
fix: comment

svmnotn commented:
Ok, I think that covers almost everything, now to find why the Eof is weird.

calebmer commented on graphql-parser/src/lexer.rs:
```diff
@@ -190,12 +190,12 @@ fn scan_string(lexer: &mut Lexer) -> (bool, bool) {
   let mut is_float = false;
   take!(lexer, '+' | '-');
-  // Integer part
+  // IntValueeger part
```
Spelling? 😉

calebmer commented on graphql-parser/src/lexer.rs:
```diff
@@ -2,14 +2,23 @@
 pub enum Token<'a> {
-  Eof,
+  Eof(usize, usize),
+  Punctuator(Punctuator, usize),
```
I really feel like `usize`-`usize` is better for consistency instead of `usize` and another `usize` inside `Spread`. It might make sense if the variables were named, but since its a tuple, its just confusing. Also, have you thought about using Rust's [`Range`](http://doc.rust-lang.org/std/ops/struct.Range.html) type?

svmnotn replied:
Range is exclusive in the upper bound so I would end up having `+ 1` all over the code

calebmer replied:
Ok, still would like to see the `Punctuator` variant as `Punctuator(Punctuator, usize, usize)` however 😊

svmnotn replied:
The issue is that the use case will end up as `Punctuator(Punctuator, start, start)` which means we have a whole variable doing nothing for most of the time, except in the one case it is spread. 

Currently it still sort of reads correctly as it is written like `Punctuator(Spread(start), end)` there is just one extra paren. I prefer it this way, but if @cksac also wants both then that will be it.

svmnotn committed:
style: change eof to only take the hi bound as the lo one is completly ignored.

As a note, eof has some interesting behavior that needs to be checked.

calebmer commented on graphql-parser/src/lexer.rs:
```diff
@@ -4,7 +4,7 @@ use std::iter::Peekable;
 #[derive(PartialEq, Debug)]
 pub enum Token<'a> {
-  Eof(usize, usize),
+  Eof(usize),
```
Does this actually fix the issue or just hide it?

svmnotn replied:
the issue is on the hi bit, so it does neither. I only removed the lo bit because it made no sense in the eof variant.

svmnotn commented on graphql-parser/src/lexer.rs:
```diff
@@ -342,7 +352,8 @@ mod tests {
   fn test_str(input: &str, expeced: &str) {
-    test_with_lex(input, Token::String(expeced));
+    let len = expeced.len();
+    test_with_lex(input, StringValue(expeced, 1, len + 1))
```
Ok, I have `len + 1` here to make it work, but WHY? This is might be related to the eof issue.

svmnotn commented:
Currently the lexer has some kind of offset when it reaches the eof. Not sure where it is added or how to fix it, as of yet.

cksac commented:
@svmnotn I am for Punctuator(Punctuator, lo, hi), as lo, hi are byte offset in the input. Not all Punctuator have 1 byte length, for example input like "@", it should be Punctuator(At, 0, 1), as &input[0..1] = "@".

for Eof token, may be we should remove the lo and hi, just leave it as Token::Eof

svmnotn committed:
feat: Punctuator has lo and hi, again

I think I finally figured out the issue, the hi bit, should be exclusive, as shown in the numbers, however when dealing with simbols it becomes inclusive.

svmnotn committed:
feat: spread `+ 1` where needed

This is the single most hackish solution. Now I'll see how numbers and strings work to see if that might lead to a clue.

svmnotn committed:
add commit so that github recognizes the previous one

svmnotn commented:
review?

calebmer commented:
Looks good to me 👍
Although I'm not the one to be reviewing the lexer code.
